### PR TITLE
feat: show update button for outdated generators

### PR DESCRIPTION
**Pull request by slavik-lvovsky**

### CODE REVIEW

File: packages/backend/src/exploregens.ts [hunk 1]
```diff
@@ -125,25 +118,26 @@ export class ExploreGens {
   }
 
   private async getFilteredGenerators(query?: string, author?: string) {
```
Comment (alex-gilin) [inline packages/backend/src/exploregens.ts:124]: consider to add a return type signature

File: packages/backend/src/exploregens.ts [hunk 2]
```diff
@@ -161,12 +155,16 @@ export class ExploreGens {
     return vscode.commands.executeCommand("yeomanUI._notifyGeneratorsChange");
   }
 
+  private setStatusBarMessage(message: string) {
```
Comment (alex-gilin) [inline packages/backend/src/exploregens.ts:162]:
```suggestion
  private setStatusBarMessage(message: string): Thenabe<string> {
```

File: packages/frontend/src/exploregens/App.vue [hunk 3]
```diff
@@ -152,26 +155,27 @@ export default {
...
-        this.rpc.invoke(action, [gen]);
-      }
+      const action = gen.action.toLowerCase();
+      this.rpc.invoke(action, [gen]);
     },
-    onQueryChange() {
-      this.debouncedGenFilterChange();
+    onQueryChange: function () {
+      //TODO: use debounce this.debouncedGenFilterChange();
+      this.getFilteredGenerators();
```
Comment (alex-gilin) [inline packages/frontend/src/exploregens/App.vue:163]: why ?

File: packages/backend/src/webSocketServer/exploregens.ts [hunk 4]
```diff
@@ -12,7 +13,8 @@ class ExploreGensWebSocketServer {
     // web socket server
     const port = process.env.PORT ? Number.parseInt(process.env.PORT) : 8082;
 
-    const wss = new Server({ port: port }, () => {
+    //const wss = new WebSocket.Server({ port: port, path: "/exploregens" }, () => {
```
Comment (alex-gilin) [inline packages/backend/src/webSocketServer/exploregens.ts:16]: remove

**Review by alex-gilin — CHANGES_REQUESTED** [review]

File: packages/backend/src/webSocketServer/exploregens.ts (hunk 4 above)
Comment (slavik-lvovsky) [reply packages/backend/src/webSocketServer/exploregens.ts:16]: Done

File: packages/backend/src/exploregens.ts (hunk 2 above)
Comment (slavik-lvovsky) [reply packages/backend/src/exploregens.ts:162]: It is not Thenable, but Disposable
Done

File: packages/backend/src/exploregens.ts (hunk 1 above)
Comment (slavik-lvovsky) [reply packages/backend/src/exploregens.ts:124]: Done

File: packages/frontend/src/exploregens/App.vue (hunk 3 above)
Comment (slavik-lvovsky) [reply packages/frontend/src/exploregens/App.vue:163]: It does not work with debounce
Need to investigate

Comment (slavik-lvovsky) [reply packages/frontend/src/exploregens/App.vue]: Fixed. Done

**Comment by slavik-lvovsky** [comment]
ready for another cycle of review

File: packages/backend/src/exploregens.ts [hunk 5]
```diff
@@ -80,7 +84,7 @@ export class ExploreGens {
         }
       }
     } catch (error) {
-      this.showAndLogError("Update Failure", error);
+      this.showAndLogError("Auto Update Failure", error);
```
Comment (alex-gilin) [inline packages/backend/src/exploregens.ts:87]: Paola? should be extended to messages ?

File: packages/backend/src/exploregens.ts [hunk 6]
```diff
@@ -107,43 +108,44 @@ export class ExploreGens {
...
     if (!_.isEmpty(gensToUpdate)) {
       this.logger.debug(messages.auto_update_started);
-      const statusBarMessage = vscode.window.setStatusBarMessage(messages.auto_update_started);
-      const promises = _.map(gensToUpdate, (genName) => this.update(genName));
+      const statusBarMessage = this.setStatusBarMessage(messages.auto_update_started);
+      const promises = _.map(gensToUpdate, (genName) => this.update(genName, true));
       const failedToUpdateGens: any[] = _.compact(await Promise.all(promises));
       if (!_.isEmpty(failedToUpdateGens)) {
         const errMessage = messages.failed_to_update_gens(failedToUpdateGens);
         this.showAndLogError("Update Failure", errMessage);
```
Comment (alex-gilin) [inline packages/backend/src/exploregens.ts:116]: Paola? should be extended to messages ?

File: packages/backend/src/exploregens.ts [hunk 7]
```diff
@@ -107,43 +108,44 @@ export class ExploreGens {
...
       }
       this.setInstalledGens();
       statusBarMessage.dispose();
-      vscode.window.setStatusBarMessage(messages.auto_update_finished, 10000);
+      this.setStatusBarMessage(messages.auto_update_finished, 10000);
```
Comment (alex-gilin) [inline packages/backend/src/exploregens.ts:120]: 10000 - magic number ?

File: packages/backend/test/exploregens.spec.ts [hunk 8]
```diff
@@ -172,42 +231,57 @@ describe("exploregens unit test", () => {
...
+          dispose: () => {
+            return;
+          },
+        });
 
       await exploregens["install"](gen);
     });
   });
 
   describe("getFilteredGenerators", () => {
```
Comment (alex-gilin) [inline packages/backend/test/exploregens.spec.ts:245]: please consider to add a negative tests

File: packages/frontend/src/exploregens/App.vue [hunk 9]
```diff
@@ -152,26 +151,28 @@ export default {
...
+      debouncer();
     },
     isInVsCode() {
       return typeof acquireVsCodeApi !== "undefined";
     },
     async getFilteredGenerators() {
       const recommended = this.recommended === ALL_GENS ? "" : this.recommended;
-      const res = await this.rpc.invoke("getFilteredGenerators", [this.query, recommended]);
-      this.gens = _.map(res[0], (gen) => {
+      console.log("query - " + this.query);
```
Comment (alex-gilin) [inline packages/frontend/src/exploregens/App.vue:166]: whether it for debugging ?

File: packages/frontend/src/exploregens/App.vue [hunk 10]
```diff
@@ -152,26 +151,28 @@ export default {
...
+      const packagesData = await this.rpc.invoke("getFilteredGenerators", [_.get(this, "query", ""), recommended]);
+      this.gens = _.map(packagesData.packages, (gen) => {
         gen.action = this.actionName(gen);
         gen.color = this.actionColor(gen);
         return gen;
       });
-      this.total = res[1];
+      this.total = packagesData.total;
+      console.log("gens - " + this.gens);
+      console.log("total - " + this.total);
```
Comment (alex-gilin) [inline packages/frontend/src/exploregens/App.vue:175]: whether it for debugging ?

**Review by alex-gilin — CHANGES_REQUESTED** [review]

File: packages/backend/src/exploregens.ts (hunk 7 above)
Comment (slavik-lvovsky) [reply packages/backend/src/exploregens.ts:120]: It was PO requirement

File: packages/backend/test/exploregens.spec.ts (hunk 8 above)
Comment (slavik-lvovsky) [reply packages/backend/test/exploregens.spec.ts:245]: for each uninstall, install and update methods there are 2 types of tests: positive and negative
One flow passes successfully and another one fails.

File: packages/frontend/src/exploregens/App.vue (hunk 9 above)
Comment (slavik-lvovsky) [reply packages/frontend/src/exploregens/App.vue:166]: Removed

File: packages/frontend/src/exploregens/App.vue (hunk 10 above)
Comment (slavik-lvovsky) [reply packages/frontend/src/exploregens/App.vue:175]: Removed

File: packages/backend/src/exploregens.ts (hunk 6 above)
Comment (slavik-lvovsky) [reply packages/backend/src/exploregens.ts:116]: Done. She already gave me new strings

File: packages/backend/src/exploregens.ts (hunk 5 above)
Comment (slavik-lvovsky) [reply packages/backend/src/exploregens.ts:87]: Done. She already gave me new strings

**Review by alex-gilin — APPROVED** [review]

**Pull request merged**